File: leetcode-solutions/ruby/747. Largest Number At Least Twice of Others.rb
# https://leetcode.com/problems/largest-number-at-least-twice-of-others/

# @param {Integer[]} nums
# @return {Integer}
def dominant_index(nums)
  large = -1 
  large_i = 0
  i_large = -1
  
  nums.each_with_index do |n, i|
      if n > large
          i_large = large
          large = n
          large_i = i
      elsif n > i_large
          i_large = n
      end
  end
  
  large >= i_large * 2 ? large_i : -1
end

File: leetcode-solutions/go/747. Largest Number At Least Twice of Others.go
// https://leetcode.com/problems/largest-number-at-least-twice-of-others/

func dominantIndex(nums []int) int {
	var largestNum int
	var ans int

	for i, n := range nums {
		if largestNum < n {
			largestNum = n
			ans = i
		}
	}

	for _, n := range nums {
		if n != largestNum && 2*n > largestNum {
			return -1
		}
	}

	return ans
}